differs/rpm_diff.go

```diff
@@ -0,0 +1,312 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package differs
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	pkgutil "github.com/GoogleCloudPlatform/container-diff/pkg/util"
+	"github.com/GoogleCloudPlatform/container-diff/util"
+	dockertar "github.com/containers/image/docker/tarfile"
+	godocker "github.com/fsouza/go-dockerclient"
+	"github.com/sirupsen/logrus"
+)
+
+type RPMAnalyzer struct {
+}
+
+// Name returns the name of the analyzer.
+func (a RPMAnalyzer) Name() string {
+	return "RPMAnalyzer"
+}
+
+// Diff compares the installed rpm packages of image1 and image2.
+func (a RPMAnalyzer) Diff(image1, image2 pkgutil.Image) (util.Result, error) {
+	diff, err := singleVersionDiff(image1, image2, a)
+	return diff, err
+}
+
+// Analyze collects information of the installed rpm packages on image.
+func (a RPMAnalyzer) Analyze(image pkgutil.Image) (util.Result, error) {
+	analysis, err := singleVersionAnalysis(image, a)
+	return analysis, err
+}
+
+// getPackages returns a map of installed rpm package on image.
+func (a RPMAnalyzer) getPackages(image pkgutil.Image) (map[string]util.PackageInfo, error) {
+	path := image.FSPath
+	packages := make(map[string]util.PackageInfo)
+	if _, err := os.Stat(path); err != nil {
+		// invalid image directory path
+		return packages, err
+	}
+
+	// try to find the rpm binary in bin/ or usr/bin/
+	rpmBinary := filepath.Join(path, "bin/rpm")
+	if _, err := os.Stat(rpmBinary); err != nil {
+		rpmBinary = filepath.Join(path, "usr/bin/rpm")
+		if _, err = os.Stat(rpmBinary); err != nil {
+			return packages, nil
+		}
+	}
+
+	return rpmDataFromContainer(image)
+}
+
+// rpmDataFromContainer runs image in a container, queries the data of
+// installed rpm packages and returns a map of packages.
+func rpmDataFromContainer(image pkgutil.Image) (map[string]util.PackageInfo, error) {
+	packages := make(map[string]util.PackageInfo)
+
+	client, err := godocker.NewClientFromEnv()
+	if err != nil {
+		return packages, err
+	}
+
+	// if image is a tar archive, we need to load it
+	imageName := image.Source
+	if pkgutil.IsTar(imageName) {
+		// NOTE: all work below is only required for multilayer tar
+		// archives generated by `docker save`.  Flat file-system
+		// archives generated by `docker create` can just be imported
+		// without potential name conflicts.  However, those are not
+		// supported yet.
+		imageName, err = generateNewArchive(image.Source)
```

Comment:
nit: use `generateNewArchive(imageName)` since you already created the variable here

Reply:
Thanks, I will fix that.